core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,312 @@
+package io.substrait.relation;
+
+import io.substrait.expression.AbstractExpressionVisitor;
+import io.substrait.expression.Expression;
+import io.substrait.expression.ExpressionVisitor;
+import io.substrait.expression.FieldReference;
+import io.substrait.expression.ImmutableFieldReference;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Class used to visit all child relations from a root relation and optionally replace subtrees by
+ * overriding a visitor method. The traversal will include relations inside of subquery expressions.
+ * By default, no subtree substitution will be performed. However, if a visit method is overridden
+ * to return a non-empty optional value, then that value will replace the relation in the tree.
+ */
+public class RelCopyOnWriteVisitor extends AbstractRelVisitor<Optional<Rel>, RuntimeException> {
+
+  private static boolean allEmpty(Optional<?>... optionals) {
+    for (Optional<?> optional : optionals) {
+      if (optional.isPresent()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static class Pointer<T> {
+    private T value;
+
+    public void set(T value) {
+      this.value = value;
+    }
+
+    public boolean wasSet() {
+      return value != null;
+    }
+  }
+
+  public interface Getter<T, U> {
+    public T get(U element);
+  }
+
+  public interface Builder<T, U> {
+    public U build(U element, T e);
+  }
+
+  public static <T> Optional<List<T>> transformList(
+      Collection<T> items, Function<T, Optional<T>> transform) {
+    return transformList(items, transform, element -> (T) element, (element, e) -> (T) e);
+  }
+
+  public static <T, U> Optional<List<U>> transformList(
+      Collection<U> items,
+      Function<T, Optional<T>> transform,
+      Getter<T, U> getter,
+      Builder<T, U> builder) {
+    var changed = new Pointer<Boolean>();
+    var newItems =
```

Comment:
This ends up creating a new list with each iteration, as opposed to using a CopyOnWriteArrayList. Do you know of a way to do this using streams so that this doesn't happen, @jacques-n?

Reply:
Upon reading a bit more, CopyOnWriteArrayList is sucky in terms of perf because it copies the underlying array every time a change is made. What I'd do instead is create a new list only if the original list is changing and update it as it is traversed (pretty trivial).